napari/utils/colormaps/colormap.py

```diff
@@ -306,48 +315,213 @@ def shuffle(self, seed: int):
+
+    def _map_precast(self, values, apply_selection) -> np.ndarray:
+        """Map values to colors.
+
+        Parameters
+        ----------
+        values : np.ndarray
+            Values to be mapped. It need to be already cast using
+            cast_labels_to_minimum_type_auto
+
+        Returns
+        -------
+        np.ndarray of shape (N, M, 4)
+            Mapped colors.
+
+        Notes
+        -----
+        it is implemented for thumbnail labels,
+        where we already have cast values
+        """
         mapped = np.zeros(values.shape + (4,), dtype=np.float32)
+        colors = self._values_mapping_to_minimum_values_set(apply_selection)[1]
         for idx in np.ndindex(values.shape):
             value = values[idx]
-            if value in self.color_dict:
-                color = self.color_dict[value]
-                if len(color) == 3:
-                    color = np.append(color, 1)
-                mapped[idx] = color
-            else:
-                mapped[idx] = self.default_color
-        # If using selected, disable all others
-        if self.use_selection:
-            mapped[~np.isclose(values, self.selection)] = 0
+            mapped[idx] = colors[value]
         return mapped
 
+    @cached_property
+    def _num_unique_colors(self) -> int:
+        """Count the number of unique colors in the colormap."""
+        return len({tuple(x) for x in self.color_dict.values()})
+
+    def _clear_cache(self):
+        super()._clear_cache()
+        if "_num_unique_colors" in self.__dict__:
+            del self.__dict__["_num_unique_colors"]
+        if "_label_mapping_and_color_dict" in self.__dict__:
+            del self.__dict__["_label_mapping_and_color_dict"]
+        if "_numpy_mapper" in self.__dict__:
+            del self.__dict__["_numpy_mapper"]
+
+    def _values_mapping_to_minimum_values_set(
+        self, apply_selection=True
+    ) -> Tuple[Dict[Optional[int], int], Dict[int, np.ndarray]]:
+        """Create mapping from original values to minimum values set.
+        To use minimum possible dtype for labels.
+
+        Returns
+        -------
+        Dict[Optional[int], int]
+            Mapping from original values to minimum values set.
+        Dict[int, np.ndarray]
+            Mapping from new values to colors.
+
+        """
+        if self.use_selection and apply_selection:
+            return {self.selection: 1, None: 0}, {
+                0: np.array((0, 0, 0, 0)),
+                1: self.color_dict.get(
+                    self.selection,
+                    self.default_color,
+                ),
+            }
+
+        return self._label_mapping_and_color_dict
+
+    @cached_property
+    def _label_mapping_and_color_dict(
+        self,
+    ) -> Tuple[Dict[Optional[int], int], Dict[int, np.ndarray]]:
+        color_to_labels: Dict[Tuple[int, ...], List[Optional[int]]] = {}
+        labels_to_new_labels: Dict[Optional[int], int] = {None: 0}
```

Comment:
Should this be `{None: DEFAULT_VALUE}`?

Reply:
I also wonder if we should call it "NONE_VALUE" or "NONE_TEXTURE_VALUE". During review I find myself constantly wondering "default for what..."